File: class11/java/TCC1101-SimuladorAP.java
// TCC1101-SimuladorAP.java | Aula 11
// Simulador de Autômato de Pilha (AP) com trace da pilha.
// Reconhece aⁿbⁿ (n>=1) por estado final.
// Espelho Java do TCC1101-SimuladorAP.py (mesmas entradas e saídas).
import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

public class TCC1101SimuladorAP {

    static class Resultado {
        boolean aceita;
        List<String> trace;
        Resultado(boolean aceita, List<String> trace) {
            this.aceita = aceita;
            this.trace = trace;
        }
    }

    static Resultado simularAP(String palavra) {
        List<String> trace = new ArrayList<>();
        Stack<String> pilha = new Stack<>();
        pilha.push("Z"); // fundo da pilha
        String estado = "q0";
        int i = 0;
        trace.add("Início: estado=" + estado + ", pilha=" + pilha + ", entrada='" + palavra + "'");

        while (i < palavra.length()) {
            char c = palavra.charAt(i);
            String topo = pilha.isEmpty() ? "ε" : pilha.peek();

            if (estado.equals("q0") && c == 'a' && topo.equals("Z")) {
                pilha.push("A");
                trace.add("δ(q0,a,Z)→(q0,AZ) | lê '" + c + "' | pilha=" + pilha);
                i++;
            } else if (estado.equals("q0") && c == 'a' && topo.equals("A")) {
                pilha.push("A");
                trace.add("δ(q0,a,A)→(q0,AA) | lê '" + c + "' | pilha=" + pilha);
                i++;
            } else if (estado.equals("q0") && c == 'b' && topo.equals("A")) {
                estado = "q1";
                trace.add("δ(q0,b,A)→(q1,A) | lê '" + c + "' | pilha=" + pilha);
                i++;
            } else if (estado.equals("q1") && c == 'b' && topo.equals("A")) {
                pilha.pop();
                trace.add("δ(q1,b,A)→(q1,ε) | lê '" + c + "' | pilha=" + pilha);
                i++;
            } else {
                trace.add("SEM transição: estado=" + estado + ", c='" + c + "', topo=" + topo + " → REJEITA");
                return new Resultado(false, trace);
            }
        }

        // Transição ε para q2
        String topo = pilha.isEmpty() ? "ε" : pilha.peek();
        if (estado.equals("q1") && topo.equals("Z")) {
            estado = "q2";
            trace.add("δ(q1,ε,Z)→(q2,Z) | lê ε | pilha=" + pilha);
        }

        boolean aceita = estado.equals("q2");
        trace.add("Fim: estado=" + estado + " → " + (aceita ? "ACEITA" : "REJEITA"));
        return new Resultado(aceita, trace);
    }

    public static void main(String[] args) {
        String[] testes = {"ab", "aabb", "aaabbb", "aab", "abb", "ba", ""};
        System.out.printf("%-12s %s%n", "Palavra", "Resultado");
        System.out.println("-".repeat(30));
        for (String w : testes) {
            Resultado r = simularAP(w);
            System.out.printf("%-12s %s%n", w.isEmpty() ? "ε" : w, r.aceita ? "ACEITA" : "REJEITA");
        }

        System.out.println("\n=== Trace detalhado para 'aabb' ===");
        Resultado r = simularAP("aabb");
        for (String linha : r.trace) {
            System.out.println(linha);
        }
    }
}

File: class11/python/TCC1101-SimuladorAP.py
#!/usr/bin/env python3
"""TCC1101-SimuladorAP.py | Aula 11
Simulador de Autômato de Pilha (AP) com trace da pilha.
Reconhece aⁿbⁿ (n>=1) por estado final.
AP: M = ({q0,q1,q2}, {a,b}, {Z,A}, δ, q0, Z, {q2})
δ(q0,a,Z) = (q0,AZ)    -- empilha A
δ(q0,a,A) = (q0,AA)    -- empilha A
δ(q0,b,A) = (q1,A)     -- troca para q1 (consome A no topo lendo b)
δ(q1,b,A) = (q1,ε)     -- desempilha A
δ(q1,ε,Z) = (q2,Z)     -- aceita (pilha só com Z)
"""
from typing import List, Tuple


def simular_ap(palavra: str) -> Tuple[bool, List[str]]:
    """Simula o AP. Retorna (aceita, trace)."""
    trace: List[str] = []
    pilha: List[str] = ["Z"]  # fundo da pilha
    estado = "q0"
    i = 0
    trace.append(f"Início: estado={estado}, pilha={pilha}, entrada='{palavra}'")

    while i < len(palavra):
        c = palavra[i]
        topo = pilha[-1] if pilha else "ε"

        if estado == "q0" and c == "a" and topo == "Z":
            pilha.append("A")
            trace.append(f"δ(q0,a,Z)→(q0,AZ) | lê '{c}' | pilha={pilha}")
            i += 1
        elif estado == "q0" and c == "a" and topo == "A":
            pilha.append("A")
            trace.append(f"δ(q0,a,A)→(q0,AA) | lê '{c}' | pilha={pilha}")
            i += 1
        elif estado == "q0" and c == "b" and topo == "A":
            estado = "q1"
            trace.append(f"δ(q0,b,A)→(q1,A) | lê '{c}' | pilha={pilha}")
            i += 1
        elif estado == "q1" and c == "b" and topo == "A":
            pilha.pop()
            trace.append(f"δ(q1,b,A)→(q1,ε) | lê '{c}' | pilha={pilha}")
            i += 1
        else:
            trace.append(f"SEM transição: estado={estado}, c='{c}', topo={topo} → REJEITA")
            return False, trace

    # Transição ε para q2
    topo = pilha[-1] if pilha else "ε"
    if estado == "q1" and topo == "Z":
        estado = "q2"
        trace.append(f"δ(q1,ε,Z)→(q2,Z) | lê ε | pilha={pilha}")

    aceita = estado == "q2"
    trace.append(f"Fim: estado={estado} → {'ACEITA' if aceita else 'REJEITA'}")
    return aceita, trace


if __name__ == "__main__":
    testes = ["ab", "aabb", "aaabbb", "aab", "abb", "ba", ""]
    print(f"{'Palavra':<12} {'Resultado'}")
    print("-" * 30)
    for w in testes:
        ok, _ = simular_ap(w)
        print(f"{w or 'ε':<12} {'ACEITA' if ok else 'REJEITA'}")

    print("\n=== Trace detalhado para 'aabb' ===")
    _, trace = simular_ap("aabb")
    for linha in trace:
        print(linha)
